feat: Adjust output position

TillerBurr opened the pull request:
Closes #101. 

This is the implementation that I came up with. Ran into an issue where I actually need to pass in at least the first character of the last line. Using the molten kernel and notebook navigator's run and move, it wouldn't show the output automatically after running the cell and moving to the next cell, as the first position wouldn't actually be in the cell. 

Most of the cell markers that I've seen (and the only one I use) is a comment (`# %%`), it made sense to me to shift up to the last line that is not a comment and non-empty. This only counts for top level comments, i.e. no indentation.

TillerBurr committed:
feat: adjust output position

IF a cell ends in empty or comment lines, move the output up to below the last line of code.

benlubas commented on rplugin/python3/molten/code_cell.py:
```diff
@@ -69,3 +69,11 @@ def get_text(self, nvim: Nvim) -> str:
+        lines = self.get_text(self.nvim).splitlines()
+        for i, line in enumerate(reversed(lines)):
+            # Show after first non-empty, non-comment line.
+            if line != '' and not line.startswith("#"):
```
Molten supports a ton of languages, so I'd really like to stay away from a hard coded comment string. Making this configurable (with a list of possible comment strings), or using treesitter + the active kernel's language to get the correct comment string is better. Probably prefer the TS method.

benlubas replied:
I think that just making this configurable would be fine, and a lot less effort

TillerBurr replied:
Done.

benlubas commented on rplugin/python3/molten/moltenbuffer.py:
```diff
@@ -325,7 +325,8 @@ def update_interface(self) -> None:
         if self.options.virt_text_output:
             for span, output in self.outputs.items():
-                output.show_virtual_output(span.end)
+                offset = span.calculate_output_offset()
```
Doing this with virtual text can cause some weird behavior, although it's not really the end of the world.

![image](https://github.com/benlubas/molten-nvim/assets/56943754/79262f93-bf11-407b-9d8b-c3564cc440b0)

TillerBurr replied:
I think I figured it out.

benlubas commented on rplugin/python3/molten/moltenbuffer.py:
```diff
@@ -395,7 +396,9 @@ def _show_selected(self, span: CodeCell) -> None:
         if self.should_show_floating_win:
-            self.outputs[span].show_floating_win(span.end)
+            offset = span.calculate_output_offset()
+            self.outputs[span].show_floating_win(span.end, offset)
```
`show_floating_win` is also called from `:MoltenShowOutput` and `:MoltenEnterOutput`.

Those don't get called with offsets, so the offset is only displayed properly right after the cell is run.

TillerBurr replied:
I moved it into `show_floating_win` and `show_virtual_output`, so no need to add additional calls.

benlubas commented on rplugin/python3/molten/outputbuffer.py:
```diff
@@ -218,11 +218,11 @@ def show_virtual_output(self, anchor: Position) -> None:
         self.canvas.present()
 
-    def show_floating_win(self, anchor: Position) -> None:
+    def show_floating_win(self, anchor: Position, offset: int = 0) -> None:
```
can we just calculate the offset in here so we don't have to always remember to calculate it before calling this function?

Or at least make it default to calculating it, that way the windows render consistently.

benlubas commented on rplugin/python3/molten/outputbuffer.py:
```diff
@@ -218,11 +218,11 @@ def show_virtual_output(self, anchor: Position) -> None:
-        win_row = self._buffer_to_window_lineno(anchor.lineno + 1)
-        if win_row == 0:  # anchor position is off screen
+        win_row = self._buffer_to_window_lineno(anchor.lineno + 1) + offset
+        if win_row <= 0:  # anchor position is off screen
```
did this cause problems? i don't think it should have, but if it did that's probably a bug or broken assumption that might break other things, I'd want to look into that

TillerBurr replied:
It didn't cause any immediate problems, but since the `+ offset` was added and offset is negative, `win_row` could then be negative and this was to catch that possibility.

benlubas reviewed (changes requested):
Big thing that 100% needs to change: this needs to be behind a configuration option.

Other large concern:
![image](https://github.com/benlubas/molten-nvim/assets/56943754/4e797f7b-738a-4ec8-8d2f-6175b6e7165a)

Where is my cursor? Not in the floating window. This is the reason that floating windows close when you move out of the cell. I think the `molten_output_virt_lines` option would mostly fix that problem if people didn't like it.

..but that doesn't work with this change currently:

![image](https://github.com/benlubas/molten-nvim/assets/56943754/a26b55d8-fceb-422b-82fe-d7a22b1d6149)

(virtual padding displays where the old window would have rendered)

I also left some comments about the code.

---
edit: another way to fix the "is my cursor in this floating window or not" problem would be to change the logic that causes a floating window to close, and make it close when it overlaps with the window in addition to when it leaves the code cell.

But, given that you're doing a lot of work to try and keep the window open after run and move is called, I don't think you will like this option.

TillerBurr commented:
Thanks for the feedback. While I am happy to implement this, I would like to get your thoughts about it. Is it too complex of a change, just to move the visual output up a few lines? If you don't think it's worth it, I can close the pull request and issue.

benlubas commented:
After thinking about this a little more, I think that what you want to do will always cause some type of "weird" behavior.

- if you want to move windows and virtual text by a fixed offset, then `:MoltenEvaluateLine` is showing output above cells
- if you want to move it up so it covers empty lines
	- virt text: then you can add to the cell below the displayed output
	- float: then your cursor can be under the floating window
	- you need to introduce the comment string logic to get exactly the behavior that you want

I still think that the "cover empty lines" solution can be implemented pretty cleanly, and none of those issues are deal breakers for merging this (b/c they're behind a configuration flag, we'll just document them).

This PR is already most of the way there, just a single bug to fix and a few configuration options to add (+ documentation which I can write).

TillerBurr committed:
Merge branch 'benlubas:main' into output-position

TillerBurr committed:
Move offset logic to use a `Position`. Add configuration options.

TillerBurr commented:
I've got something that is more in line with what you were thinking. I changed the logic to be based on position, rather than on a code cell. That makes way more sense to me.

There was one thing that I couldn't figure out. ~~I~~ You had the idea to use virtual lines, so it shouldn't cover many lines, if any. In my testing, it only covers the last line: `# %%`, but it's shifted purposefully to cover that up. The issue is that on initial render, it appears to be shifted sometimes:
![Screenshot 2024-01-30 133155](https://github.com/benlubas/molten-nvim/assets/49296311/ba45b9b7-92c3-4c74-888e-ad875d7f2ab3)
Leaving and re-entering the cell leads to the expected render.
![Screenshot 2024-01-30 133215](https://github.com/benlubas/molten-nvim/assets/49296311/0a0086f6-e609-49ca-af02-d09f80555f4a)

Is there somewhere that I'm missing that does the initial render only?

TillerBurr committed:
Fix bad config defaults

benlubas commented on rplugin/python3/molten/outputbuffer.py:
```diff
@@ -218,11 +218,38 @@ def show_virtual_output(self, anchor: Position) -> None:
+                False,
+            )[0]
+
+            if current_line != "" and not current_line.startswith(self.options.comment_string):
```
I'm having problems when the comment string isn't set bc the default value is `""` and all strings start with `""`.  Solution is probably to have it default to None.

But I think a better solution is to make it a list, and have it default to empty. That way if people who're using this plugin for more than one notebook language, they wouldn't have to change the configuration around.

<img width="918" alt="image" src="https://github.com/benlubas/molten-nvim/assets/56943754/f2b0a2b8-b5ba-4d49-85e4-a55fa2451574">

TillerBurr replied:
Wow. Thanks, can't believe I made that mistake. Of course every line starts with an empty string. I forgot to change my settings and remove that config option when I was testing it. I changed it and renamed the option to more accurately reflect what it does, it wasn't super clear what it was used for otherwise.

benlubas reviewed (changes requested):
Small bug when testing. I'll also dig into the virtual text issue you mentioned.

benlubas commented:
I'm seeing the opposite behavior to what you describe. I see the window initially render correctly and then after I scroll, the window goes back to the position it would be in if this change just wasn't here at all.

TillerBurr committed:
Comment Strings, Don't create `DynamicPosition`
Change `comment_string` to a `cover_lines_starting_with`, since it could
be for multiple languages.
Creating a `DynamicPosition` seemed unnecessary. It was only being used
to get the line number. Changed to just decrement the `lineno`
directly.

TillerBurr commented:
I have no idea why it's shifting and why it would be so different between the two machines. 

Could it be something between versions of neovim? I'm testing on `v0.9.5`. 

It's broken on `NVIM v0.10.0-dev-2241+gbe5cf3383`. I get the same error message as https://github.com/neovim/neovim/issues/27211 but can't seem to get a dump. Everything works, except when it tries to add in the virtual lines: https://github.com/TillerBurr/molten-nvim/blob/2b19c471d8cec9596c0fa2ebe2b17c01c059c52b/rplugin/python3/molten/outputbuffer.py#L351

Even if that wasn't trying to add the virtual text, it would still crash if `output_virt_lines` was true.

benlubas commented:
I'm also on 9.5. I'm pretty sure a couple parts of this plugin break on nightly :/

Will get around to fixing that at some point.

TillerBurr commented:
An update: Been pretty busy lately. I still can't get this to work properly. I'm pretty sure it deals with the `_buffer_to_window_lineno` and setting the `win_row`. The `shape` keeps changing, but the offset remains the same. Now, I just have to figure out the proper offset based on the window row. 

https://github.com/benlubas/molten-nvim/assets/49296311/205e824c-3569-43d8-8cf3-0564a7d82e6c

I was logging a bunch of stuff and was able to confirm that's the case.
[molten.log](https://github.com/benlubas/molten-nvim/files/14316652/molten.log)

benlubas commented:
Here's a more in depth explanation of what I'm seeing, as well as the reason I think it's happening.

<img width="438" alt="image" src="https://github.com/benlubas/molten-nvim/assets/56943754/dbca327a-499e-416e-82a2-8917c680115f">

- Run the code with 2 blank lines below it
- The window shows up in the right spot, the virtual text shows up under the window which pushes the main buffer contents down so I can still see everything

<img width="438" alt="image" src="https://github.com/benlubas/molten-nvim/assets/56943754/2489f237-78f7-4a7e-9578-dd9b2ce723e1">

- I scroll (`<c-e>`)
- The window repositions underneath the virtual text, but still 2 lines higher than the end of the cell (so the offset is still taking effect)

<img width="438" alt="image" src="https://github.com/benlubas/molten-nvim/assets/56943754/561ee36c-a989-4d17-ac95-45a51ca49590">

---

Here's what I think is happening. The original location of the window is calculated based on the bottom line of the cell. The position is calculated correctly. Then the window is shown and the virtual text is added to the screen.

When I scroll, the bottom line of the cell is used again to calculate the window position, but now the bottom line is below the virtual text. So the position is "correctly" calculated as `old_position + 4` where there are 4 lines of virtual text.

The solution is to subtract the number of empty lines from the line number instead of from the screen position, if that makes sense?

TillerBurr commented:
Makes sense. I was thinking along the same line. I didn't realize that the anchor line number stayed the same but the window position changed. Obvious in hindsight.

Weird that we are seeing different behavior. That I don't understand.

benlubas commented:
Yeah I still have no idea why we're seeing different things. Hopefully the fix for me is the fix for you too lol

TillerBurr committed:
adjust window position based on virtual text

TillerBurr committed:
offset before window position, not after

TillerBurr commented:
Alright, I think I have it. At least, I'm seeing the desired behavior. Hopefully you do too.

benlubas commented:
@TillerBurr sorry this took me so long. The window positioning appears to behave itself. If there's nothing else you want to change I can merge this

TillerBurr commented:
No problem at all, I get busy too. 

The only thing would be the name of the config variable. It was called 'cover' before the virtual text was added. I can't think of a better name at the moment. Unless you have any ideas, it should be good to merge.

benlubas commented:
I think the current names are fine. I'll merge this tomorrow morning
